File: project_euler/Problem097/Problem097.java
package project_euler;

public class Problem097 {
    public void solution() {
        int target = 7830457;
        int factor = 28433;
        int[] product = new int[2];
        long[] auxiliary = new long[2];
        getPower(target, product, auxiliary);
        getProduct(factor, 1, product, auxiliary);

        long result = (long) product[1] * (long) 1e5 + product[0];
        System.out.println(result);
    }

    private void getPower(int value, int[] product, long[] auxiliary) {
        if (value < 8) {
            product[0] = (int) Math.pow(2, value);
            product[1] = 0;
            return;
        }

        getPower(value >> 1, product, auxiliary);
        getSquare(product, auxiliary);
        if (0 != (value & 1)) {
            getDouble(product);
        }
    }

    private void getProduct(int factor, int value, int[] product, long[] auxiliary) {
        auxiliary[0] = (long) product[0] * factor + value;
        auxiliary[1] = (long) product[1] * factor;
        carry(product, auxiliary);
    }

    private void getSquare(int[] product, long[] auxiliary) {
        auxiliary[0] = (long) product[0] * product[0];
        auxiliary[1] = (long) product[0] * product[1] << 1;
        carry(product, auxiliary);
    }

    private void getDouble(int[] product) {
        product[0] <<= 1;
        product[1] <<= 1;
        if (product[0] >= (int) 1e5) {
            product[1] += product[0] / (int) 1e5;
            product[0] %= (int) 1e5;
        }
        product[1] %= (int) 1e5;
    }

    private void carry(int[] product, long[] auxiliary) {
        if (auxiliary[0] >= (long) 1e5) {
            auxiliary[1] += auxiliary[0] / (long) 1e5;
            auxiliary[0] %= (long) 1e5;
        }
        auxiliary[1] %= (long) 1e5;
        for (int i = 0; i < 2; ++i) {
            product[i] = (int) auxiliary[i];
        }
    }
}

File: project_euler/Problem097/Problem097.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem097() {
    target := 7830457
    factor := 28433
    product := make([]int, 2)
    auxiliary := make([]int64, 2)
    getPower(target, product, auxiliary)
    getProduct(factor, 1, product, auxiliary)

    result := int64(product[1]) * int64(1e5) + int64(product[0])
    fmt.Println(result)
}

func getPower(value int, product []int, auxiliary []int64) {
    if value < 8 {
        product[0] = int(math.Pow(2, float64(value)))
        product[1] = 0
        return
    }

    getPower(value >> 1, product, auxiliary)
    getSquare(product, auxiliary)
    if 0 != (value & 1) {
        getDouble(product)
    }
}

func getProduct(factor int, value int, product []int, auxiliary []int64) {
    auxiliary[0] = int64(product[0]) * int64(factor) + int64(value)
    auxiliary[1] = int64(product[1]) * int64(factor)
    carry(product, auxiliary)
}

func getSquare(product []int, auxiliary []int64) {
    auxiliary[0] = int64(product[0]) * int64(product[0])
    auxiliary[1] = int64(product[0]) * int64(product[1]) << 1
    carry(product, auxiliary)
}

func getDouble(product []int) {
    product[0] <<= 1
    product[1] <<= 1
    if product[0] >= int(1e5) {
        product[1] += product[0] / int(1e5)
        product[0] %= int(1e5)
    }
    product[1] %= int(1e5)
}

func carry(product []int, auxiliary []int64) {
    if auxiliary[0] >= int64(1e5) {
        auxiliary[1] += auxiliary[0] / int64(1e5)
        auxiliary[0] %= int64(1e5)
    }
    auxiliary[1] %= int64(1e5)
    for i := 0; i < 2; i++ {
        product[i] = int(auxiliary[i])
    }
}
